File: typescript/chapters/ch12/src/core/loop.ts
import { resolve } from "node:path";

import { HookContractError, HookRegistry } from "./hooks.js";
import type { HookResult } from "./hooks.js";
import {
  isChatMessage,
  systemMessage,
  toolMessage,
  userMessage,
  validateToolPairing,
} from "./messages.js";
import type { ChatMessage, ToolCall } from "./messages.js";
import type { ModelClient, ModelReply, ModelRequest } from "./model.js";
import { PermissionDecision, PermissionPolicy, PermissionRequest } from "./permissions.js";
import type { PreparedToolCall, ToolContext, ToolRegistry, ToolResult } from "./tools.js";
import { copyToolResult, toolError } from "./tools.js";

// AgentRunner 是 Agent Loop：在模型、权限与工具前后接入 Hook/处理器，并保证每个 tool call 都有回填结果。
export class AgentRunError extends Error {
  override readonly name: string = "AgentRunError";
}

export class AgentLimitError extends AgentRunError {
  override readonly name: string = "AgentLimitError";
}

export class IncompleteModelReplyError extends AgentRunError {
  override readonly name: string = "IncompleteModelReplyError";
}

export interface RunResult {
  readonly finalText: string;
  readonly history: readonly ChatMessage[];
  readonly turns: number;
}

export interface AgentRunnerOptions {
  readonly model: ModelClient;
  readonly tools: ToolRegistry;
  readonly systemPrompt: string;
  // 动态提示 Provider 可替换固定 systemPrompt；每次模型请求前都会重新调用 render()。
  readonly systemPromptProvider?: SystemPromptProvider;
  readonly workspace: string;
  readonly maxTurns?: number;
  readonly identity?: string;
  readonly permissionPolicy?: PermissionPolicy;
  readonly hooks?: HookRegistry;
  readonly toolRoundObserver?: ToolRoundObserver;
  // 请求处理器只影响下一次模型请求；工具结果处理器必须在回填前完成。
  readonly historyProcessor?: RequestHistoryProcessor;
  readonly toolResultProcessor?: ToolResultProcessor;
  readonly turnLifecycle?: TurnLifecycle;
  readonly modelRequestExecutor?: ModelRequestExecutor;
}

// 处理器产出必须仍满足消息配对，确保压缩不破坏模型协议。
export interface RequestHistoryProcessor {
  prepare(history: readonly ChatMessage[]): Promise<readonly ChatMessage[]>;
}

// 动态提示 Provider 对 AgentRunner 保持零参数契约，内部数据源由组合根在构建时绑定。
export interface SystemPromptProvider {
  // 每轮模型调用临时读取动态系统提示，不把渲染结果写入对话历史。
  render(): string;
}

export interface ModelRequestExecutor {
  // Executor 接管模型调用时必须显式 beginTurn，确保恢复状态与 Agent 回合对齐。
  beginTurn(): void;
  complete(request: ModelRequest): Promise<ModelReply>;
}

// TurnLifecycle 是记忆等跨请求能力的边界：beginTurn 在首轮前执行，
// beforeModel 只给下一次模型请求附加上下文，complete 在最终结果返回前收尾。
export interface TurnLifecycle {
  beginTurn(query: string): Promise<void>;
  beforeModel(): readonly ChatMessage[];
  complete(history: readonly ChatMessage[]): Promise<void>;
}

export type ToolResultProcessor = (
  results: readonly ToolResult[],
) => Promise<readonly ToolResult[]> | readonly ToolResult[];

// 观察器可在模型请求前提供指导，并在工具轮完成后更新内部状态。
export interface ToolRoundObserver {
  beforeModel(): readonly ChatMessage[];
  recordToolRound(toolNames: readonly string[]): void;
}

interface ToolExecution {
  readonly result: ToolResult;
  readonly additionalContext: readonly ChatMessage[];
  readonly preventContinuation: boolean;
}

export class AgentRunner {
  readonly #model: ModelClient;
  readonly #tools: ToolRegistry;
  readonly #systemPrompt: string;
  readonly #systemPromptProvider: SystemPromptProvider | undefined;
  readonly #workspace: string;
  readonly #maxTurns: number;
  readonly #identity: string;
  readonly #permissionPolicy: PermissionPolicy | undefined;
  readonly #hooks: HookRegistry;
  readonly #toolRoundObserver: ToolRoundObserver | undefined;
  readonly #historyProcessor: RequestHistoryProcessor | undefined;
  readonly #toolResultProcessor: ToolResultProcessor | undefined;
  readonly #turnLifecycle: TurnLifecycle | undefined;
  // P11 执行器接管模型请求；不注入时保持旧路径直接调用 ModelClient。
  readonly #modelRequestExecutor: ModelRequestExecutor | undefined;
  readonly #history: ChatMessage[] = [];

  constructor(options: AgentRunnerOptions) {
    const maxTurns = options.maxTurns === undefined ? 20 : options.maxTurns;
    if (!Number.isInteger(maxTurns) || maxTurns <= 0) {
      throw new Error("maxTurns must be a positive integer");
    }
    const identity = options.identity === undefined ? "user" : options.identity;
    if (identity.trim().length === 0) {
      throw new Error("identity must not be empty");
    }
    if (options.systemPrompt.trim().length === 0) {
      throw new Error("systemPrompt must not be empty");
    }
    if (
      options.systemPromptProvider !== undefined &&
      typeof options.systemPromptProvider.render !== "function"
    ) {
      // provider 契约在构造时校验，避免运行到模型调用阶段才发现缺 render()。
      throw new TypeError("systemPromptProvider must implement render()");
    }
    if (
      options.modelRequestExecutor !== undefined &&
      (typeof options.modelRequestExecutor.beginTurn !== "function" ||
        typeof options.modelRequestExecutor.complete !== "function")
    ) {
      // 执行器契约也在构造时校验，避免运行到模型调用阶段才发现缺方法。
      throw new TypeError("modelRequestExecutor must implement beginTurn() and complete()");
    }

    this.#model = options.model;
    this.#tools = options.tools;
    this.#systemPrompt = options.systemPrompt;
    this.#systemPromptProvider = options.systemPromptProvider;
    this.#workspace = resolve(options.workspace);
    this.#maxTurns = maxTurns;
    this.#identity = identity;
    this.#permissionPolicy =
      options.permissionPolicy === undefined && options.hooks !== undefined
        ? new PermissionPolicy()
        : options.permissionPolicy;
    this.#hooks = options.hooks === undefined ? new HookRegistry() : options.hooks;
    this.#toolRoundObserver = options.toolRoundObserver;
    this.#historyProcessor = options.historyProcessor;
    this.#toolResultProcessor = options.toolResultProcessor;
    this.#turnLifecycle = options.turnLifecycle;
    this.#modelRequestExecutor = options.modelRequestExecutor;
  }

  get history(): readonly ChatMessage[] {
    return Object.freeze([...this.#history]);
  }

  async run(prompt: string): Promise<RunResult> {
    // canonical history 始终完整追加；请求级压缩或生命周期注入只影响每次模型快照。
    const submitted = userMessage(prompt);
    const promptHook = await this.#hooks.runUserPrompt(submitted);
    this.#history.push(submitted, ...promptHook.additionalContext);
    if (this.#turnLifecycle !== undefined) {
      // 生命周期先于首轮模型请求执行；选择失败由实现方降级，不能阻塞主 Agent。
      await this.#turnLifecycle.beginTurn(prompt);
    }
    // 执行器与 memory lifecycle 对齐回合边界；P01-P10 不注入时继续直接调用 ModelClient。
    this.#modelRequestExecutor?.beginTurn();
    const context: ToolContext = Object.freeze({
      workspace: this.#workspace,
      identity: this.#identity,
    });

    let stopHookActive = false;
    for (let turn = 1; turn <= this.#maxTurns; turn += 1) {
      validateToolPairing(this.#history);
      const preparedHistory: unknown =
        this.#historyProcessor === undefined
          ? this.#history
          : await this.#historyProcessor.prepare(Object.freeze([...this.#history]));
      if (
        !Array.isArray(preparedHistory) ||
        !preparedHistory.every((message: unknown) => isChatMessage(message))
      ) {
        throw new AgentRunError("Request history processor returned invalid messages");
      }
      const requestHistory = Object.freeze([...preparedHistory]);
      // requestHistory 只决定模型本次能看到什么，canonical history 仍保持完整追加。
      validateToolPairing(requestHistory);
      // beforeModel 注入只进入本次模型请求，不写入 canonical history。
      const turnGuidance =
        this.#turnLifecycle === undefined ? [] : this.#turnLifecycle.beforeModel();
      if (
        !Array.isArray(turnGuidance) ||
        !turnGuidance.every((message: unknown) => isChatMessage(message))
      ) {
        throw new AgentRunError("Turn lifecycle returned invalid model guidance");
      }
      const observerGuidance =
        this.#toolRoundObserver === undefined ? [] : this.#toolRoundObserver.beforeModel();
      // 观察器指导是请求级内容，只影响当次模型输入；TODO 提醒等不写入持久历史。
      // 系统提示在每轮组装请求前渲染，使工具列表和选中记忆等运行态变化下一轮生效。
      const systemPrompt = this.#renderSystemPrompt();
      const requestMessages = Object.freeze([
        systemMessage(systemPrompt),
        ...requestHistory,
        ...turnGuidance,
        ...observerGuidance,
      ]);
      validateToolPairing(requestMessages);
      const tools = this.#tools.snapshot();
      const request: ModelRequest = Object.freeze({
        messages: requestMessages,
        tools: tools.openAITools(),
      });
      // 模型边界一次只发送当前轮请求；canonical history 继续由本类维护。
      const reply = await (this.#modelRequestExecutor === undefined
        ? this.#model.complete(request)
        : this.#modelRequestExecutor.complete(request));
      // 有执行器时，内部重试、压缩和 fallback 都在 complete() 内收束，不会重新进入外层循环。

      if (reply.finishReason === "length") {
        throw new IncompleteModelReplyError("Model output reached the token limit");
      }
      if (reply.finishReason === "content_filter") {
        throw new AgentRunError("Model response was blocked by the content filter");
      }

      const assistant = reply.message;
      this.#history.push(assistant);
      if (assistant.toolCalls.length === 0) {
        if (assistant.content === null) {
          throw new AgentRunError("Model stopped without final text or tool calls");
        }
        const stopHook = await this.#hooks.runStop(this.#history, stopHookActive);
        if (stopHook.forceContinue !== undefined) {
          this.#history.push(...stopHook.additionalContext, stopHook.forceContinue);
          stopHookActive = true;
          continue;
        }
        return await this.#complete(assistant.content, turn);
      }

      // 一条 assistant 消息中的每个调用都必须回填一次，错误和拒绝也不例外。
      const results: ToolResult[] = [];
      const deferredContext: ChatMessage[] = [];
      let stoppedResultIndex: number | undefined;
      for (const call of assistant.toolCalls) {
        let result: ToolResult;
        if (stoppedResultIndex !== undefined) {
          result = toolError(
            "hook_stopped_continuation",
            "Skipped after PostToolUse requested a stop",
          );
        } else {
          const execution = await this.#executeTool(call, context, tools);
          result = execution.result;
          deferredContext.push(...execution.additionalContext);
          if (execution.preventContinuation) {
            stoppedResultIndex = results.length;
          }
        }
        results.push(result);
      }
      const processedResults = await this.#processToolResults(results, assistant.toolCalls);
      for (const [index, call] of assistant.toolCalls.entries()) {
        const result = processedResults[index];
        if (result === undefined) {
          throw new AgentRunError("Tool execution did not produce a paired result");
        }
        // 每个工具调用都必须回填一个 tool message，否则后续模型请求会破坏协议配对。
        this.#history.push(toolMessage(result.content, call.id));
      }
      if (this.#toolRoundObserver !== undefined) {
        this.#toolRoundObserver.recordToolRound(assistant.toolCalls.map((call) => call.name));
      }
      this.#history.push(...deferredContext);
      if (stoppedResultIndex !== undefined) {
        const stoppedResult = processedResults[stoppedResultIndex];
        if (stoppedResult === undefined) {
          throw new AgentRunError("PostToolUse stop did not preserve its result");
        }
        return await this.#complete(stoppedResult.content, turn);
      }
    }

    throw new AgentLimitError(`Agent exceeded maxTurns=${this.#maxTurns}`);
  }

  async #processToolResults(
    results: readonly ToolResult[],
    calls: readonly ToolCall[],
  ): Promise<readonly ToolResult[]> {
    // 没有处理器时也深拷贝结果，避免工具内部可变对象污染 canonical history。
    if (this.#toolResultProcessor === undefined) {
      return Object.freeze(results.map((result) => copyToolResult(result)));
    }
    try {
      // 处理器收到只读快照；返回批次必须与工具调用数一致，否则整批按错误回填。
      const input = Object.freeze(results.map((result) => copyToolResult(result)));
      const processed: unknown = await this.#toolResultProcessor(input);
      if (!Array.isArray(processed) || processed.length !== calls.length) {
        throw new Error("tool result processor returned an invalid batch");
      }
      return Object.freeze(processed.map((result) => copyToolResult(result)));
    } catch {
      // 处理器失败时整批返回受控错误，仍与原工具调用数量一一对应。
      return Object.freeze(
        calls.map(() => toolError("tool_result_processing_error", "Tool result processing failed")),
      );
    }
  }

  // 每轮模型请求前渲染系统提示；provider 缺失时使用构建期固定字符串。
  #renderSystemPrompt(): string {
    // Provider 输出是模型边界输入，空值立即失败而不是回退到旧缓存。
    const rendered =
      this.#systemPromptProvider === undefined
        ? this.#systemPrompt
        : this.#systemPromptProvider.render();
    if (typeof rendered !== "string" || rendered.trim().length === 0) {
      throw new AgentRunError("System prompt provider returned an empty prompt");
    }
    return rendered;
  }

  async #executeTool(
    call: ToolCall,
    context: ToolContext,
    tools: ToolRegistry,
  ): Promise<ToolExecution> {
    // 工具执行边界统一产出 ToolResult，Hook、权限和 handler 异常都不可见原始堆栈。
    let prepared: PreparedToolCall;
    try {
      prepared = tools.prepare(call);
    } catch {
      return execution(toolError("tool_preparation_error", "Tool preparation failed"));
    }
    if (prepared.error !== undefined) {
      return execution(prepared.error);
    }

    let preHook: HookResult;
    try {
      preHook = await this.#hooks.runPreTool(prepared);
    } catch (error) {
      return execution(
        error instanceof HookContractError
          ? toolError("hook_contract_error", "PreToolUse hook returned an invalid update")
          : toolError("hook_execution_error", "PreToolUse hook failed"),
      );
    }
    const effective = preHook.updatedInput === undefined ? prepared : preHook.updatedInput;
    // Pre Hook 可阻断或建议权限，实际拒绝仍由统一权限策略裁决。
    if (preHook.blockingError !== undefined) {
      return execution(preHook.blockingError, preHook.additionalContext);
    }

    if (this.#permissionPolicy !== undefined) {
      // Hook 只提交结构化建议；系统 deny 仍在同一策略合并中拥有最高优先级。
      try {
        const decision = await this.#permissionPolicy.decide(
          new PermissionRequest({
            prepared: effective,
            context,
            recommendations: hookRecommendations(preHook),
          }),
        );
        if (!decision.isAllowed) {
          return execution(decision.toToolResult(), preHook.additionalContext);
        }
      } catch {
        return execution(
          toolError("permission_evaluation_error", "Permission evaluation failed"),
          preHook.additionalContext,
        );
      }
    }

    let result = await tools.invoke(effective, context);
    let postHook: HookResult;
    try {
      postHook = await this.#hooks.runPostTool(effective, result);
    } catch {
      return execution(
        toolError("hook_execution_error", "PostToolUse hook failed"),
        preHook.additionalContext,
      );
    }
    if (postHook.updatedOutput !== undefined) {
      result = postHook.updatedOutput;
    }
    return execution(
      result,
      [...preHook.additionalContext, ...postHook.additionalContext],
      postHook.preventContinuation,
    );
  }

  async #complete(finalText: string, turns: number): Promise<RunResult> {
    validateToolPairing(this.#history);
    // complete 使用完整 canonical history，让 extractor 看到请求级压缩前的原始会话。
    if (this.#turnLifecycle !== undefined) {
      await this.#turnLifecycle.complete(Object.freeze([...this.#history]));
    }
    return Object.freeze({
      finalText,
      history: Object.freeze([...this.#history]),
      turns,
    });
  }
}

function execution(
  result: ToolResult,
  additionalContext: readonly ChatMessage[] = [],
  preventContinuation = false,
): ToolExecution {
  return Object.freeze({
    result,
    additionalContext: Object.freeze([...additionalContext]),
    preventContinuation,
  });
}

function hookRecommendations(hook: HookResult): readonly PermissionDecision[] {
  if (hook.permissionBehavior === "passthrough") {
    return [];
  }
  return [
    new PermissionDecision(
      hook.permissionBehavior,
      `PreToolUse hook requested ${hook.permissionBehavior}`,
      "pre-tool-hook",
    ),
  ];
}

File: python/ch02_agent/agent_ch02/core/loop.py
"""Agent 核心循环。

这是什么：Agent 运行时的核心控制逻辑，负责循环调用模型和执行工具
Java 类比：类似 @Service class AgentService，编排模型调用、工具执行和消息流转
为什么需要：实现 Agent 的核心控制流，连接模型推理和工具执行，管理对话状态

Java 角度：这是应用服务，不负责创建 OpenAI 客户端或 PowerShell 进程；
它只编排 ModelClient、ToolRegistry 和消息历史。
"""

from dataclasses import dataclass
from pathlib import Path
from typing import Protocol

from .messages import ChatMessage, system_message, tool_message, user_message, validate_tool_pairing
from .model import ModelClient, ModelRequest
from .tools import PreparedToolCall, ToolContext, ToolRegistry, ToolResult, tool_error


class AgentRunError(Exception):
    """Agent 执行过程中的领域错误。

    这是什么：Agent 运行时异常的基类
    Java 类比：类似 AgentExecutionException
    为什么需要：统一 Agent 运行时错误，便于上层捕获和处理
    """


class AgentLimitError(AgentRunError):
    """达到最大模型调用轮数。

    这是什么：循环次数限制异常
    Java 类比：类似 MaxIterationsExceededException
    为什么需要：防止死循环，保护资源和成本
    """


class IncompleteModelReplyError(AgentRunError):
    """模型输出因 token 限制被截断。

    这是什么：模型输出不完整异常
    Java 类比：类似 IncompleteLLMResponseException
    为什么需要：标识模型输出被截断，提示需要增加 max_tokens 或简化任务
    """


@dataclass(frozen=True, slots=True)
class ToolAuthorizationDecision:
    """授权结果：是否允许，以及给模型看的原因。

    这是什么：工具授权决策的值对象
    Java 类比：类似 record AuthorizationDecision(boolean allowed, String reason)
    为什么需要：封装授权结果，让模型能理解为什么被拒绝
    """
    allowed: bool  # True 表示可以执行工具；False 表示只生成拒绝结果。
    reason: str  # 给人和模型看的原因，拒绝时尤其重要。


class ToolAuthorizer(Protocol):
    """工具授权接口，类似 Java 中的鉴权 Service。

    这是什么：工具授权器的接口定义
    Java 类比：interface ToolAuthorizer { AuthorizationDecision authorize(PreparedToolCall call, ToolContext ctx); }
    为什么需要：定义授权契约，支持不同授权策略（自动批准、人工审核、策略规则）
    """
    def authorize(self, prepared: PreparedToolCall, context: ToolContext) -> ToolAuthorizationDecision: ...


@dataclass(frozen=True, slots=True)
class RunResult:
    """一次 Agent 运行结束后返回给调用方的结果。

    这是什么：Agent 运行结果的值对象
    Java 类比：类似 record RunResult(String finalText, List<ChatMessage> history, int turns)
    为什么需要：封装运行结果，提供完整的对话历史和元数据
    """
    final_text: str  # 模型最后一次返回的普通文本。
    history: tuple[ChatMessage, ...]  # 完整对话副本，用于测试和审计。
    turns: int  # 实际调用模型的次数，从 1 开始。


class AgentRunner:
    """一轮轮调用模型，直到模型返回最终文本。

    这是什么：Agent 循环控制器，编排模型调用、工具执行和消息管理
    Java 类比：类似 @Service class AgentOrchestrator
    为什么需要：实现 Agent 的 ReAct 循环（Reasoning + Acting），管理状态和流程

    这是全章最重要的类。可以把它看成一个 Service：它只负责编排，
    不负责 HTTP 请求细节，也不负责 PowerShell 进程细节。
    """

    def __init__(
        self,
        model: ModelClient,
        tools: ToolRegistry,
        system_prompt: str,
        workspace: str,
        max_turns: int = 20,
        identity: str = "user",
        authorizer: ToolAuthorizer | None = None,
    ) -> None:
        """初始化 Agent 运行器。

        这是什么：构造器，通过依赖注入接收所有外部依赖
        Java 类比：类似 @Autowired 构造器注入依赖
        为什么需要：解耦 Agent 逻辑和外部实现，便于测试和配置切换
        """
        # 构造函数只保存依赖和长期配置。和 Java 构造器注入一样，依赖从外部传进来。
        if max_turns <= 0:  # 验证最大轮数必须为正数
            raise ValueError("max_turns 必须是正整数")
        if not identity.strip():  # 验证身份标识不能为空
            raise ValueError("identity 不能为空")
        if not system_prompt.strip():  # 验证系统提示不能为空
            raise ValueError("system_prompt 不能为空")
        self._model = model  # 模型接口：可以是真实 DeepSeek，也可以是测试 Fake。
        self._tools = tools  # 工具注册表：保存模型可以调用的"手"。
        self._system_prompt = system_prompt  # 每轮都要放在消息最前面的系统约束。
        self._workspace = str(Path(workspace).resolve())  # 工具允许使用的工作目录。
        self._max_turns = max_turns  # 单次任务最多调用模型多少次，防止死循环。
        self._identity = identity  # 当前调用者身份，后续权限系统会使用。
        self._authorizer = authorizer  # 可选授权器；没有时通常用于离线测试。
        self._history: list[ChatMessage] = []  # 不包含 system prompt 的可变会话历史。

    @property
    def history(self) -> tuple[ChatMessage, ...]:
        """返回对话历史的不可变副本。

        这是什么：历史记录的只读访问器
        Java 类比：类似 public List<ChatMessage> getHistory() { return List.copyOf(history); }
        为什么需要：保护内部状态，防止外部代码修改历史记录
        """
        # tuple 是不可变序列。返回副本，避免外部代码直接修改 Agent 内部历史。
        return tuple(self._history)

    def run(self, prompt: str) -> RunResult:
        """执行 Agent 主循环，直到模型返回最终答案或达到轮数限制。

        这是什么：Agent 的核心运行方法，实现 ReAct 循环
        Java 类比：类似 public RunResult run(String prompt) throws AgentRunError
        为什么需要：启动 Agent 循环，处理用户请求，返回最终结果
        """
        # 第一步：把用户问题放入历史。system prompt 不放这里，它每轮请求时临时加在最前面。
        self._history.append(user_message(prompt))  # 添加用户消息到历史

        # 所有工具共享同一份工作目录和调用者身份，工具自己不能随意决定这些边界。
        context = ToolContext(self._workspace, self._identity)  # 构造工具执行上下文

        # 一次循环就是一次模型请求。设置上限是为了防止模型无限要求调用工具。
        for turn in range(1, self._max_turns + 1):  # 循环调用模型，从 1 开始计数
            # 在花钱请求模型前，先确认上一轮工具调用都有结果。
            validate_tool_pairing(self._history)  # 校验工具调用和结果是否配对

            # 快照保证"模型看到的工具"和"这一轮真正能执行的工具"完全一致。
            snapshot = self._tools.snapshot()  # 获取工具注册表快照

            # Python 的 *self._history 类似 Java 中把一个 List 展开放进新 List。
            request = ModelRequest(
                messages=(system_message(self._system_prompt), *self._history),  # 组装消息列表，system prompt 放最前面
                tools=snapshot.openai_tools(),  # 添加工具定义
            )

            # 核心循环只调用接口，不知道内部是 DeepSeek、OpenAI 还是假模型。
            reply = self._model.complete(request)  # 调用模型获取响应

            # length 表示回答被 token 上限截断，不能把半截内容当成最终答案。
            if reply.finish_reason == "length":  # 输出因长度限制被截断
                raise IncompleteModelReplyError("模型输出达到 token 上限，回答不完整")
            if reply.finish_reason == "content_filter":  # 输出被内容过滤器拦截
                raise AgentRunError("模型回答被内容过滤器拦截")

            # 模型消息要先入历史，后面的 tool 消息才能通过 call.id 与它配对。
            assistant = reply.message  # 提取 assistant 消息
            self._history.append(assistant)  # 添加到历史记录

            # 没有工具调用，表示模型认为任务完成了，此时退出循环。
            if not assistant.tool_calls:  # 模型未请求工具调用
                if assistant.content is None:  # 但也没有返回文本内容
                    raise AgentRunError("模型已停止，但没有返回最终文本或工具调用")
                validate_tool_pairing(self._history)  # 最后验证历史完整性
                return RunResult(assistant.content, tuple(self._history), turn)  # 返回最终结果

            for call in assistant.tool_calls:  # 遍历所有工具调用请求
                # prepare 只校验，不执行 PowerShell。
                prepared = snapshot.prepare(call)  # 准备工具调用（仅校验，不执行）
                result: ToolResult  # 声明结果变量类型
                if prepared.error is not None:  # 准备阶段发现错误（如工具不存在、参数非法）
                    # 未知工具、错误 JSON 等失败已经变成可回填的 ToolResult。
                    result = prepared.error  # 使用准备阶段的错误结果
                elif self._authorizer is not None:  # 有授权器时需要人工批准
                    try:
                        # 有授权器时，必须先获得明确允许，才能进入 invoke。
                        decision = self._authorizer.authorize(prepared, context)  # 请求授权
                        if not decision.reason.strip():  # 授权结果必须包含原因
                            raise ValueError("工具授权结果必须说明原因")
                        result = snapshot.invoke(prepared, context) if decision.allowed else tool_error("permission_denied", decision.reason)  # 根据授权结果执行或拒绝
                    # 授权器属于外部边界，无论抛出哪种异常都必须默认拒绝。
                    except Exception:  # noqa: BLE001 | 捕获所有授权器异常
                        # 授权系统本身报错时默认拒绝，这叫 fail-closed。
                        result = tool_error("permission_denied", "工具授权过程发生异常，已按默认拒绝处理")  # 默认拒绝策略
                else:  # 无授权器时直接执行（通常用于测试）
                    # 单元测试通常不注入授权器，直接使用 Fake 工具执行器。
                    result = snapshot.invoke(prepared, context)  # 直接执行工具

                # 无论成功、失败还是拒绝，每个 call.id 都必须写回一条 tool 消息。
                self._history.append(tool_message(result.content, call.id))  # 添加工具结果到历史

        # 循环用完仍未得到最终文本，说明模型一直在调用工具。
        raise AgentLimitError(f"Agent 已达到最大模型调用轮数 max_turns={self._max_turns}")  # 达到最大轮数限制
